lib/workflow/stores/mongo.js

```diff
@@ -643,18 +643,34 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
             graphId: data.graphId,
             reachable: true
         };
-        query['dependencies.' + data.taskId] = {
-            $in: [data.state, Constants.Task.States.Finished]
-        };
         var update = {
             $unset: {}
         };
-        update.$unset['dependencies.' + data.taskId] = '';
         var options = {
             multi: true
         };
+        
+        query['dependencies.' + data.taskId] = {
+            $in: [data.state, Constants.Task.States.Finished]
+        };
 
-        return waterline.taskdependencies.updateMongo(query, update, options);
+        return waterline.taskdependencies.find(query).then(function(res) {
+           if (_.isEmpty(res)) {
+                // no task found under dependencies try search key anyOf
+                delete query['dependencies.' + data.taskId];
+                query['dependencies.anyOf.' + data.taskId] = {
+                    $in: [data.state, Constants.Task.States.Finished]
+                };
+
+                update.$unset['dependencies.anyOf'] = '';
+                
+                return waterline.taskdependencies.updateMongo(query, update, options);
+            }
+            else {
```

Comment:
If a task appears in both `dependencies.` and `dependencies.any`, the `if/else` logic will limit to only handle one case. 

For example:
```javascript
{
  dependencies: {
    '123': 'succeed',
    '456': 'finished',
    'anyOf': {
      '123': 'timeout',
      '789': 'finished'
     }
  }
}
```

The task `123` appears two times in the dependencies logic, if it is finished, then both placed should be updated.